### ISSUE TITLE
Possibly unnecessary duplicate bindings test

### ISSUE THREAD

**Issue by johnedquinn** [partiql/partiql-lang-kotlin#859]
## Description
- While writing the implementation for aggregations (logical GROUP BY) (see #794), I came across a test (`groupByDuplicateAliasesTest`) in `EvaluatingCompilerGroupByTest`.
- Here's a summary of the test:
```sql
--- TEST # 1:
SELECT *
FROM suppliers AS s
GROUP BY s.supplierId AS dup, s.supplierName as dup;
--- EXPECTS:
<< { 'dup': 10, 'dup': 'Umbrella' }, { 'dup': 11, 'dup': 'Initech' } >>

----------------------------------------------------------------------------------

--- TEST # 2:
SELECT dup
FROM suppliers AS s
GROUP BY s.supplierId AS dup, s.supplierName as dup;
--- EXPECTS:
<< { 'dup': 10 }, { 'dup': 11 } >>
```
- The input, `suppliers`, equals:
```sql
[
  { supplierId: 10, supplierName: "Umbrella" },
  { supplierId: 11, supplierName: "Initech" }
]
```
- To me, TEST # 2 and the expected output make sense. We grab the first binding `dup` seen.
- But, TEST # 1, to me, is a bit confusing -- and it doesn't provide any benefit to the end-user.

## To Reproduce
- See above.

## Expected Behavior
- To me, TEST # 1 should throw an exception due to user-defined duplicate bindings.

## Additional Context
- Java Version: 11 (irrelevant though)
- PartiQL version: 0.8.1-SNAPSHOT

## More Context
- The main reason I wanted to bring this up is because of how it affected my implementation of aggregations.
- Essentially, the `EvaluatingCompiler` uses globally unique generated group key IDs (see `UniqueNameMeta`, `IsGroupAttributeMeta` and `IsSyntheticNameMeta`) -- seemingly to handle both variable scoping and duplicate bindings.
- So, with the `LogicalResolved` plan, I was hoping that I wouldn't have to use any of the mentioned Metas above -- as they can be messy, and several AST passes are reliant on their presence.
- So, my first implementation didn't use the metas -- I just used the GROUP KEY aliases that are explicitly or implicitly defined
- Then, via the `LogicalToLogicalResolvedVisitorTransform`, any references to the GROUP KEY aliases are resolved (without needing to use the metas to mimic globally-unique IDs).
- However, with this implementation, I couldn't pass the test mentioned above (TEST # 1).
- I ended up discussing with @RCHowell about this -- and he recommended attempting to add a parameter to the LOGICAL ID and VAR_DECL to make them unique. So, I locally implemented something to accomplish this.
- This local implementation essentially adds a `represents` parameter -- which is the logical expression that the ID is supposed to represent. In the above TEST # 1, the VAR_DECL for the first `dup` would look something like: `var_decl('dup', path(id('s'), 'supplierId'))`
- With this in place, we can logically resolve this exact scenario. And it works locally -- but it's just a fairly decently-sized change for a use-case that does not provide the user any benefit.
- Anyways, I discussed with @am357, @jpschorr, and @yliuuuu -- and it requires further discussion. Hence, this issue.

**Comment by almann** [comment]
I do not think we need to support duplicate variable bindings (so I suspect both tests are wrong).  Such behavior is not really required by the specification and I don't think it is required by the SQL Standard.


```sql
postgres=# CREATE TABLE FOO(a INT);
CREATE TABLE

postgres=# INSERT INTO FOO VALUES (1), (2), (3);
INSERT 0 3

postgres=# SELECT * FROM FOO as f, FOO as f;
2022-10-28 16:51:14.115 UTC [71] ERROR:  table name "f" specified more than once
2022-10-28 16:51:14.115 UTC [71] STATEMENT:  SELECT * FROM FOO AS f, FOO AS f;
ERROR:  table name "f" specified more than once
```

This is not quite the same as your example, but we don't have `GROUP BY ... AS ...` in SQL to compare to.

**Comment by johnedquinn** [comment]
**Note**: This comment is a bit long -- and more so to just put my thoughts on paper.

So, we may not provide as much flexibility as we'd like with our implementation. I tested [DB Fiddle](https://www.db-fiddle.com/f/4jyoMCicNSZpjMt4jFYoz5/6176) (Postgres) for the following query:
```sql
SELECT * FROM t1, t2 GROUP BY t1.a, t2.a
```

And I get a relation with two columns (namely `a` and `a`). So, while this isn't totally relevant to this specific issue (it's more relevant to what I want to do with the aggregate operator) -- I'll just leave some notes.

Let's do more partiql-specific queries:
```sql
-- # 1
SELECT * FROM t1, t2 GROUP BY t1.a, t2.a AS a
-- Rewritten:
SELECT $partiql_group_key_0 AS a, a FROM t1, t2 GROUP BY t1.a AS $partiql_group_key_0, t2.a AS a

-- # 2
SELECT * FROM t1, t2 GROUP BY t1.a AS a, t2.a AS a
-- Throw an error

-- # 3
SELECT a FROM t1, t2 GROUP BY t1.a AS a, t2.a AS a
-- Throw error

-- # 4
SELECT a FROM t1, t2 GROUP BY t1.a, t2.a
-- Unsure about this one, but Postgres throws an error (column reference "a" is ambiguous). Need to look at spec.
-- I'm leaning towards grabbing the first one -- but tabling this until further research.

-- # 5
SELECT * FROM t1, t2 GROUP BY t1.a AS a, t2.a AS b
-- Rewritten
SELECT a, b FROM t1, t2 GROUP BY t1.a AS a, t2.a AS b

-- # 6
SELECT * FROM t1, t2 GROUP BY t1.a, t2.a
-- Rewritten
SELECT $partiql_group_key_0 AS a, $partiql_group_key_1 AS b FROM t1, t2 GROUP BY t1.a AS $partiql_group_key_0, t2.a AS $partiql_group_key_1
```

I believe this can be done without metas and by performing this during AST normalization. Specifically, the visitor transform can collect **explicitly defined** group key aliases and **implicitly defined** group key aliases. When two explicitly defined group key aliases are the same -- we can throw an exception.

Implicitly defined aliases can store the typical alias conversion (`t.a -> a`) and also the alias used for evaluation (`$partiql_group_key_${number}`). That way, we can rewrite the projection list if required.

Now that I think about it, we might be able to support the duplicate bindings using this method if we wanted to.

We'd just need to be sure cover the following scenarios:
- Projection list contains group key references:
  - `SELECT someKey AS outputKey FROM ...` -> `SELECT ${theEvaluationAlias} AS outputKey FROM ...`
  - `SELECT someKey + 1 AS outputKey FROM ...` -> `SELECT ${theEvaluationAlias} + 1 AS outputKey FROM ...`
  - `SELECT someKey + 1 FROM ...` -> `SELECT ${theEvaluationAlias} + 1 AS _1 FROM ...`
  - `SELECT someKey FROM ...` -> `SELECT ${theEvaluationAlias} AS someKey FROM ...`
  - `SELECT * FROM ...` -> `SELECT ${theEvaluationAlias} AS ${theImplicitAlias}`
- Having clause rewrites to use the ${evaluationAlias}
- Order by clause rewrites to use the ${evaluationAlias}
